File: web/src/pag-image.ts
import { AlphaType, ColorType, PAGScaleMode } from './types';
import { NativeImage } from './core/native-image';
import { wasmAwaitRewind, wasmAsyncMethod, destroyVerify } from './utils/decorators';
import { PAGModule } from './binding';
import { Matrix } from './core/matrix';

@destroyVerify
@wasmAwaitRewind
export class PAGImage {
  /**
   * Create pag image from image file.
   */
  @wasmAsyncMethod
  public static async fromFile(data: File): Promise<PAGImage> {
    return new Promise((resolve, reject) => {
      const image = new Image();
      image.onload = async () => {
        resolve(PAGImage.fromSource(image));
      };
      image.onerror = (error) => {
        reject(error);
      };
      image.src = URL.createObjectURL(data);
    });
  }
  /**
   * Create pag image from image source or video source.
   * Make sure the target pixel is shown on the screen.
   * Like
   * ``` javascript
   * Image.onload = async () => {
   *   return await PAGImage.fromSource(Image)
   * }
   * ```
   */
  public static fromSource(source: TexImageSource): PAGImage {
    const nativeImage = new NativeImage(source);
    const wasmIns = PAGModule._PAGImage._FromNativeImage(nativeImage);
    if (!wasmIns) throw new Error('Make PAGImage from source fail!');
    return new PAGImage(wasmIns);
  }
  /**
   *  Creates a PAGImage object from an array of pixel data, return null if it's not valid pixels.
   */
  public static fromPixels(
    pixels: Uint8Array,
    width: number,
    height: number,
    colorType: ColorType,
    alphaType: AlphaType,
  ): PAGImage {
    const rowBytes = width * (colorType === ColorType.ALPHA_8 ? 1 : 4);
    const dataPtr = PAGModule._malloc(pixels.byteLength);
    const dataOnHeap = new Uint8Array(PAGModule.HEAPU8.buffer, dataPtr, pixels.byteLength);
    dataOnHeap.set(pixels);
    const wasmIns = PAGModule._PAGImage._FromPixels(dataPtr, width, height, rowBytes, colorType, alphaType);
    if (!wasmIns) throw new Error('Make PAGImage from pixels fail!');
    return new PAGImage(wasmIns);
  }
  /**
   * Creates a PAGImage object from the specified backend texture, return null if the texture is
   * invalid.
   */
  public static fromTexture(textureID: number, width: number, height: number, flipY: boolean) {
    const wasmIns = PAGModule._PAGImage._FromTexture(textureID, width, height, flipY);
    if (!wasmIns) throw new Error('Make PAGImage from texture fail!');
    return new PAGImage(wasmIns);
  }

  public wasmIns;
  public isDestroyed = false;

  public constructor(wasmIns: any) {
    this.wasmIns = wasmIns;
  }
  /**
   * Returns the width in pixels.
   */
  public width(): number {
    return this.wasmIns._width() as number;
  }
  /**
   * Returns the height in pixels.
   */
  public height(): number {
    return this.wasmIns._height() as number;
  }
  /**
   * Returns the current scale mode. The default value is PAGScaleMode::LetterBox.
   */
  public scaleMode(): PAGScaleMode {
    return this.wasmIns._scaleMode() as PAGScaleMode;
  }
  /**
   * Specifies the rule of how to scale the content to fit the image layer's original size.
   * The matrix changes when this method is called.
   */
  public setScaleMode(scaleMode: PAGScaleMode) {
    this.wasmIns._setScaleMode(scaleMode);
  }
  /**
   * Returns a copy of current matrix.
   */
  public matrix(): Matrix {
    const wasmIns = this.wasmIns._matrix();
    if (!wasmIns) throw new Error('Get matrix fail!');
    return new Matrix(wasmIns);
  }
  /**
   * Set the transformation which will be applied to the content.
   * The scaleMode property will be set to PAGScaleMode::None when this method is called.
   */
  public setMatrix(matrix: Matrix) {
    this.wasmIns._setMatrix(matrix.wasmIns);
  }
  /**
   * Destroy the pag image.
   */
  public destroy(): void {
    this.wasmIns.delete();
    this.isDestroyed = true;
  }
}

File: android/libpag/src/main/java/org/libpag/PAGImage.java
package org.libpag;

import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.SurfaceTexture;
import android.opengl.EGLContext;
import android.os.Build;
import android.util.Log;

import org.extra.tools.LibraryLoadUtils;

public class PAGImage {
    /**
     * Creates a PAGImage object from an android.graphics.Bitmap object, return null if it's not
     * valid Bitmap.
     */
    public static PAGImage FromBitmap(Bitmap bitmap) {
        if (bitmap != null && bitmap.getConfig() == Bitmap.Config.ARGB_4444) {
            bitmap = bitmap.copy(Bitmap.Config.ARGB_8888, false);
        }
        long context = LoadFromBitmap(bitmap);
        if (context == 0) {
            return null;
        }
        return new PAGImage(context);
    }

    /**
     * Creates a PAGImage object from a path of a image file, return null if the file does not exist
     * or it's not a valid image file.
     */
    public static PAGImage FromPath(String path) {
        long context = LoadFromPath(path);
        if (context == 0) {
            return null;
        }
        return new PAGImage(context);
    }

    /**
     * Creates a PAGImage object from the specified byte data, return null if the bytes is empty or it's not a valid
     * image file.
     */
    public static PAGImage FromBytes(byte[] bytes) {
        if (bytes == null || bytes.length == 0) {
            return null;
        }

        long context = LoadFromBytes(bytes, bytes.length);
        if (context == 0) {
            return null;
        }
        return new PAGImage(context);
    }

    /***
     * Load a pag file from assets, returns null if the file does not exist or the data is not a
     * pag file.
     */
    public static PAGImage FromAssets(AssetManager manager, String fileName) {
        long context = LoadFromAssets(manager, fileName);
        if (context == 0) {
            return null;
        }
        return new PAGImage(context);
    }

    /**
     * Creates a PAGImage object from the specified OpenGL texture ID, returns null if the texture
     * is invalid or there is no current openGL context. Note:
     * 1. The caller must make sure that the texture outlive the lifetime of the returned PAGImage
     * object.
     * 2. The current openGL context should be the same as the shareContext parameter of
     * {@link PAGSurface#FromSurfaceTexture(SurfaceTexture, EGLContext)} or the context which PAGSurface.FromTexture() uses,
     * so that the returned PAGImage object can be drawn on to that PAGSurface.
     * 3. The caller can use fence sync objects to synchronise texture content (see {@link
     * PAGPlayer#flushAndFenceSync(long[])} and {@link PAGPlayer#waitSync(long)}).
     *
     * @param textureID   the id of the texture.
     * @param textureTarget either GL_TEXTURE_EXTERNAL_OES or GL_TEXTURE_2D.
     * @param width       the width of the texture.
     * @param height      the height of the texture.
     *
     * @see PAGSurface#FromSurfaceTexture(SurfaceTexture, EGLContext)
     * @see PAGSurface#FromTexture(int, int, int)
     * @see PAGSurface#FromTexture(int, int, int, boolean)
     * @see PAGSurface#FromTextureForAsyncThread(int, int, int)
     * @see PAGSurface#FromTextureForAsyncThread(int, int, int, boolean)
     * @see PAGPlayer#flushAndFenceSync(long[])
     * @see PAGPlayer#waitSync(long)
     */
    public static PAGImage FromTexture(int textureID, int textureTarget, int width, int height) {
        return FromTexture(textureID, textureTarget, width, height, false);
    }

    /**
     * Creates a PAGImage object from the specified OpenGL texture ID, returns null if the texture
     * is invalid or there is no current openGL context. Note:
     * 1. The caller must make sure that the texture outlive the lifetime of the returned PAGImage
     * object.
     * 2. The current openGL context should be the same as the shareContext parameter of
     * {@link PAGSurface#FromSurfaceTexture(SurfaceTexture, EGLContext)} or the context which PAGSurface.FromTexture() uses,
     * so that the returned PAGImage object can be drawn on to that PAGSurface.
     * 3. The caller can use fence sync objects to synchronise texture content (see {@link
     * PAGPlayer#flushAndFenceSync(long[])} and {@link PAGPlayer#waitSync(long)}).
     *
     * @param textureID   the id of the texture.
     * @param textureTarget either GL_TEXTURE_EXTERNAL_OES or GL_TEXTURE_2D.
     * @param width       the width of the texture.
     * @param height      the height of the texture.
     * @param flipY indicates the origin of the texture, true means 'bottom-left', false means 'top-left'.
     *
     * @see PAGSurface#FromSurfaceTexture(SurfaceTexture, EGLContext)
     * @see PAGSurface#FromTexture(int, int, int)
     * @see PAGSurface#FromTexture(int, int, int, boolean)
     * @see PAGSurface#FromTextureForAsyncThread(int, int, int)
     * @see PAGSurface#FromTextureForAsyncThread(int, int, int, boolean)
     * @see PAGPlayer#flushAndFenceSync(long[])
     * @see PAGPlayer#waitSync(long)
     */
    public static PAGImage FromTexture(int textureID, int textureTarget, int width, int height, boolean flipY) {
        long context = LoadFromTexture(textureID, textureTarget, width, height, flipY);
        if (context == 0) {
            return null;
        }
        return new PAGImage(context);
    }


    private static native long LoadFromBitmap(Bitmap bitmap);

    private static native long LoadFromPath(String path);

    private static native long LoadFromBytes(byte[] bytes, int length);

    private static native long LoadFromAssets(AssetManager manager, String fileName);

    private static native long LoadFromTexture(int textureID, int textureType,
                                               int width, int height, boolean flipY);

    PAGImage(long nativeContext) {
        this.nativeContext = nativeContext;
    }

    /**
     * Returns the width in pixels of the image.
     */
    public native int width();

    /**
     * Returns the height in pixels of the image.
     */
    public native int height();

    /**
     * Returns the current scale mode.
     */
    public native int scaleMode();

    /**
     * Specifies the rule of how to scale the image content to fit the original image size. The matrix
     * changes when this method is called.
     */
    public native void setScaleMode(int mode);

    /**
     * Returns a copy of current matrix.
     */
    public Matrix matrix() {
        float[] values = new float[9];
        nativeGetMatrix(values);
        Matrix matrix = new Matrix();
        matrix.setValues(values);
        return matrix;
    }

    private native void nativeGetMatrix(float[] values);

    /**
     * Set the transformation which will be applied to the image content. The scaleMode property
     * will be set to PAGScaleMode::None when this method is called.
     */
    public void setMatrix(Matrix matrix) {
        float[] values = new float[9];
        matrix.getValues(values);
        nativeSetMatrix(values[0], values[3], values[1], values[4], values[2], values[5]);
    }

    private native void nativeSetMatrix(float a, float b, float c,
                                        float d, float tx, float ty);

    /**
     * Free up resources used by the PAGImage instance immediately instead of relying on the
     * garbage collector to do this for you at some point in the future.
     */
    public void release() {
        nativeRelease();
    }

    private native final void nativeRelease();

    private static native final void nativeInit();

    protected void finalize() {
        nativeFinalize();
    }

    private native void nativeFinalize();

    static {
        LibraryLoadUtils.loadLibrary("pag");
        nativeInit();
    }

    long nativeContext = 0;
}
